File: Zeze/Raft/RocksRaft/CollList1.cs

using System;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Zeze.Raft.RocksRaft
{
	public class CollList1<E> : CollList<E>
	{
		public override E this[int index]
		{
			get => Data[index];
			set
			{
				if (value == null)
					throw new ArgumentNullException("value is null");

				if (this.IsManaged)
				{
					var txn = Transaction.Current;
					var log = (LogList1<E>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
					log.SetItem(index, value);
				}
				else
				{
					_list = _list.SetItem(index, value);
				}
			}
		}

		public override void Add(E item)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Add(item);
			}
			else
            {
				_list = _list.Add(item);
			}
		}

		public override void AddRange(IEnumerable<E> items)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
						Parent.ObjectId + VariableId, CreateLogBean);
				listLog.AddRange(items);
			}
			else
            {
				_list = _list.AddRange(items);
			}
		}

		public override void Clear()
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
						Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Clear();
			}
			else
            {
				_list = ImmutableList<E>.Empty;
			}
		}

		public override void Insert(int index, E item)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
					Parent.ObjectId + VariableId, CreateLogBean);
				listLog.Insert(index, item);
			}
			else
            {
				_list = _list.Insert(index, item);
			}
		}

		public override bool Remove(E item)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
						Parent.ObjectId + VariableId, CreateLogBean);
				return listLog.Remove(item);
			}
			else
            {
				var tmp = _list.Remove(item);
				if (tmp == _list)
					return false;
				_list = tmp;
				return true;
			}
		}

		public override void RemoveAt(int index)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
						Parent.ObjectId + VariableId, CreateLogBean);
				listLog.RemoveAt(index);
			}
			else
            {
				_list = _list.RemoveAt(index);
			}
		}

		public override void RemoveRange(int index, int count)
		{
			if (IsManaged)
			{
				var listLog = (LogList1<E>)Transaction.Current.LogGetOrAdd(
						Parent.ObjectId + VariableId, CreateLogBean);
				listLog.RemoveRange(index, count);
			}
			else
            {
				_list = _list.RemoveRange(index, count);
			}
		}

		public override LogBean CreateLogBean() {
			var log = new LogList1<E>();
			log.Belong = Parent;
			log.This = this;
			log.VariableId = VariableId;
			log.Value = _list;
			return log;
		}

		public override void FollowerApply(Log _log)
		{
			var log = (LogList1<E>)_log;
			foreach (var opLog in log.OpLogs)
			{
				switch (opLog.op)
				{
					case LogList1<E>.OpLog.OP_MODIFY:
						_list = _list.SetItem(opLog.index, opLog.value);
						break;
					case LogList1<E>.OpLog.OP_ADD:
						_list = _list.Insert(opLog.index, opLog.value);
						break;
					case LogList1<E>.OpLog.OP_REMOVE:
						_list = _list.RemoveAt(opLog.index);
						break;
					case LogList1<E>.OpLog.OP_CLEAR:
						_list = ImmutableList<E>.Empty;
						break;
				}
			}
		}

		public override void LeaderApplyNoRecursive(Log _log)
		{
			_list = ((LogList1<E>)_log).Value;
		}

		protected override void InitChildrenRootInfo(Record.RootInfo root)
		{
		}

		public override Bean CopyBean()
		{
			var copy = new CollList1<E>();
			copy._list = _list;
			return copy;
		}
	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Raft/RocksRaft/CollList1.java
package Zeze.Raft.RocksRaft;

import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Serialize.SerializeHelper;
import Zeze.Util.Reflect;
import org.pcollections.Empty;

public class CollList1<V> extends CollList<V> {
	protected final SerializeHelper.CodecFuncs<V> valueCodecFuncs;
	private final int logTypeId;

	public CollList1(Class<V> valueClass) {
		valueCodecFuncs = SerializeHelper.createCodec(valueClass);
		logTypeId = Zeze.Transaction.Bean.hash32("Zeze.Raft.RocksRaft.LogList1<" + Reflect.getStableName(valueClass) + '>');
	}

	private CollList1(int logTypeId, SerializeHelper.CodecFuncs<V> valueCodecFuncs) {
		this.valueCodecFuncs = valueCodecFuncs;
		this.logTypeId = logTypeId;
	}

	@Override
	public boolean add(V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			listLog.add(item);
		} else
			list = list.plus(item);
		return true;
	}

	@Override
	public boolean remove(V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			return listLog.remove(item);
		}
		var newList = list.minus(item);
		if (newList == list)
			return false;
		list = newList;
		return true;
	}

	@Override
	public void clear() {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			listLog.clear();
		} else
			list = org.pcollections.Empty.vector();
	}

	@Override
	public V set(int index, V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			return listLog.Set(index, item);
		}
		var old = list.get(index);
		list = list.with(index, item);
		return old;
	}

	@Override
	public void add(int index, V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			listLog.add(index, item);
		} else
			list = list.plus(index, item);
	}

	@Override
	public V remove(int index) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var listLog = (LogList1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			return listLog.remove(index);
		}
		var old = list.get(index);
		list = list.minus(index);
		return old;
	}

	@Override
	public LogBean createLogBean() {
		var log = new LogList1<>(logTypeId, valueCodecFuncs);
		log.setBelong(parent());
		log.setThis(this);
		log.setVariableId(variableId());
		log.setValue(list);
		return log;
	}

	@Override
	public void followerApply(Log _log) {
		@SuppressWarnings("unchecked")
		var log = (LogList1<V>)_log;
		for (var opLog : log.getOpLogs()) {
			switch (opLog.op) {
			case LogList1.OpLog.OP_MODIFY:
				list = list.with(opLog.index, opLog.value);
				break;
			case LogList1.OpLog.OP_ADD:
				list = list.plus(opLog.index, opLog.value);
				break;
			case LogList1.OpLog.OP_REMOVE:
				list = list.minus(opLog.index);
				break;
			case LogList1.OpLog.OP_CLEAR:
				list = Empty.vector();
				break;
			}
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public void leaderApplyNoRecursive(Log _log) {
		list = ((LogList1<V>)_log).getValue();
	}

	@Override
	protected void initChildrenRootInfo(Record.RootInfo root) {
	}

	@Override
	public CollList1<V> copy() {
		var copy = new CollList1<>(logTypeId, valueCodecFuncs);
		copy.list = list;
		return copy;
	}

	@Override
	public void encode(ByteBuffer bb) {
		var tmp = getList();
		bb.WriteUInt(tmp.size());
		var encoder = valueCodecFuncs.encoder;
		for (var e : tmp)
			encoder.accept(bb, e);
	}

	@Override
	public void decode(IByteBuffer bb) {
		clear();
		var decoder = valueCodecFuncs.decoder;
		for (int i = bb.ReadUInt(); i > 0; i--)
			add(decoder.apply(bb));
	}
}
